File: finalyearproject/repos/dkd-fraud/src/fraud/ids.py
"""Canonical IDs — UUID v7 bodies per the DM ID conventions; fraud's ordering key is the DID."""

from __future__ import annotations

import os
import time

DID_PREFIX = "did:dokandar:"


def uuid7() -> str:
    ms = int(time.time() * 1000)
    rand = bytearray(os.urandom(16))
    rand[0] = (ms >> 40) & 0xFF
    rand[1] = (ms >> 32) & 0xFF
    rand[2] = (ms >> 24) & 0xFF
    rand[3] = (ms >> 16) & 0xFF
    rand[4] = (ms >> 8) & 0xFF
    rand[5] = ms & 0xFF
    rand[6] = (rand[6] & 0x0F) | 0x70
    rand[8] = (rand[8] & 0x3F) | 0x80
    h = rand.hex()
    return f"{h[0:8]}-{h[8:12]}-{h[12:16]}-{h[16:20]}-{h[20:32]}"


def new_event_id() -> str:
    return uuid7()


def new_signal_id() -> str:
    return "FSG-" + uuid7()


def is_did(s: object) -> bool:
    return isinstance(s, str) and s.startswith(DID_PREFIX) and len(s) > len(DID_PREFIX)


def now_ms() -> int:
    return int(time.time() * 1000)

File: finalyearproject/repos/dkd-b2c-order/src/domain/ids.ts
// Canonical IDs (UUID v7 bodies per Domain-Model ID conventions).
import { randomBytes } from "node:crypto";

export function uuid7(unixMs = Date.now()): string {
  const b = randomBytes(16);
  b[0] = Number((BigInt(unixMs) >> 40n) & 0xffn);
  b[1] = Number((BigInt(unixMs) >> 32n) & 0xffn);
  b[2] = (unixMs >>> 24) & 0xff;
  b[3] = (unixMs >>> 16) & 0xff;
  b[4] = (unixMs >>> 8) & 0xff;
  b[5] = unixMs & 0xff;
  b[6] = (b[6] & 0x0f) | 0x70; // version 7
  b[8] = (b[8] & 0x3f) | 0x80; // variant 10
  const h = b.toString("hex");
  return `${h.slice(0, 8)}-${h.slice(8, 12)}-${h.slice(12, 16)}-${h.slice(16, 20)}-${h.slice(20)}`;
}

export const newOrd = (): string => `ORD-${uuid7()}`;
export const newEventId = (): string => uuid7();

export const isDid = (s: unknown): s is string =>
  typeof s === "string" && s.startsWith("did:dokandar:") && s.length > "did:dokandar:".length;
export const isGpid = (s: unknown): s is string =>
  typeof s === "string" && s.startsWith("GP-") && s.length > 3;
export const isOrd = (s: unknown): s is string =>
  typeof s === "string" && s.startsWith("ORD-") && s.length > 4;
